ocaml/typing/env.ml

```diff
@@ -3339,10 +3446,18 @@ let lookup_modtype ~errors ~use ~loc lid env =
   path, Subst.Lazy.force_modtype_decl mt
 
 let lookup_class ~errors ~use ~loc lid env =
-  match lid with
-  | Lident s -> lookup_ident_class ~errors ~use ~loc s env
-  | Ldot(l, s) -> lookup_dot_class ~errors ~use ~loc l s env
-  | Lapply _ -> assert false
+  let path, locks, cld =
+    match lid with
+    | Lident s -> lookup_ident_class ~errors ~use ~loc s env
+    | Ldot(l, s) -> lookup_dot_class ~errors ~use ~loc l s env
+    | Lapply _ -> assert false
+  in
+  if use then begin
+    let clda_mode = Mode.Value.legacy |> Mode.Value.disallow_right in
+    let vmode = walk_locks ~errors ~loc ~env ~item:Class ~lid clda_mode None locks in
+    submode_legacy ~errors ~loc ~env ~item:Class ~lid vmode
```

Comment:
I think it would be better to constrain the mode at the call sites of `lookup_class`.